staging/src/k8s.io/client-go/tools/auth/exec/exec.go

```diff
@@ -0,0 +1,150 @@
+
+func init() {
+	metav1.AddToGroupVersion(scheme, schema.GroupVersion{Version: "v1"})
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	utilruntime.Must(v1beta1.AddToScheme(scheme))
+	utilruntime.Must(clientauthentication.AddToScheme(scheme))
+}
+
+// LoadExecCredentialFromEnv is a helper-wrapper around LoadExecCredential that loads from the
+// well-known KUBERNETES_EXEC_INFO environment variable.
+//
+// When the KUBERNETES_EXEC_INFO environment variable is not set or is empty, then this function
+// will immediately return an error.
+func LoadExecCredentialFromEnv() (runtime.Object, *rest.Config, error) {
+	env := os.Getenv(execInfoEnv)
+	if env == "" {
+		return nil, nil, errors.New("KUBERNETES_EXEC_INFO env var is unset or empty")
+	}
+	return LoadExecCredential([]byte(env))
+}
+
+// LoadExecCredential loads the configuration needed for an exec plugin to communicate with a
+// cluster.
+//
+// LoadExecCredential expects the provided data to be a serialized client.authentication.k8s.io
+// ExecCredential object (of any version). If the provided data is invalid (i.e., it cannot be
+// unmarshalled into any known client.authentication.k8s.io ExecCredential version), an error will
+// be returned. A successfully unmarshalled ExecCredential will be returned as the first return
+// value.
+//
+// If the provided data is successfully unmarshalled, but it does not contain cluster information
+// (i.e., ExecCredential.Spec.Cluster == nil), then the returned rest.Config and error will be nil.
+//
+// Note that the returned rest.Config will use anonymous authentication, since the exec plugin has
+// not returned credentials for this cluster yet.
+func LoadExecCredential(data []byte) (runtime.Object, *rest.Config, error) {
+	obj, gvk, err := codecs.UniversalDeserializer().Decode(data, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("decode: %w", err)
+	}
+
+	expectedGK := schema.GroupKind{
+		Group: clientauthentication.SchemeGroupVersion.Group,
+		Kind:  "ExecCredential",
+	}
+	if gvk.GroupKind() != expectedGK {
+		return nil, nil, fmt.Errorf(
+			"invalid group/kind: wanted %s, got %s",
+			expectedGK.String(),
+			gvk.GroupKind().String(),
+		)
+	}
+
+	// Explicitly convert object here so that we can return a nicer error message above for when the
+	// data represents an invalid type.
+	var execCredential clientauthentication.ExecCredential
+	if err := scheme.Convert(obj, &execCredential, nil); err != nil {
+		return nil, nil, fmt.Errorf("cannot convert to ExecCredential: %w", err)
+	}
+
+	if execCredential.Spec.Cluster == nil {
+		return nil, nil, errors.New("exec credential does not contain cluster information")
+	}
+
+	restConfig, err := execClusterToConfig(execCredential.Spec.Cluster)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot create config: %w", err)
+	}
+
+	return obj, restConfig, nil
+}
+
+func execClusterToConfig(cluster *clientauthentication.Cluster) (*rest.Config, error) {
```

Comment:
going all the way through a mock kubeconfig struct seems pretty roundabout. why are we not translating these fields directly? we have the following methods which should round-trip (modulo loading CA file and resolving the proxy URL):

* execClusterToConfig: exec cluster -> rest.Config
* ConfigToExecCluster: rest.Config -> exec cluster

I expected both of these to live in similar places and have a test verifying they could round-trip with each other

Reply:
> going all the way through a mock kubeconfig struct seems pretty roundabout. why are we not translating these fields directly?

Yup, totally agree, it seems silly. I started to write a manual conversion from `exec cluster` to `rest.Config`, but then I realized I was rewriting a lot of kubeconfig loading logic. To avoid code duplication, and possibly missing out on future `rest.Config`-creation features/bug fixes, I elected to create the mock kubeconfig. This is certainly a loosely held opinion, though.

> I expected both of these to live in similar places

Agreed. This was another "where should these code elements go" moment. If we put `ConfigToExecCluster`, per your suggestion above[0], in `k8s.io/client-go/plugin/pkg/client/auth/exec`, then I guess it would follow existing dependency graph edges, now that I think of it. I thought that `plugin` packages were only supposed to be imported into `main` packages, and the like, for registration. However, I'm just now realizing that `k8s.io/client-go/rest` depends on `k8s.io/client-go/plugin/pkg/client/auth/exec`, and `k8s.io/client-go/tools/auth/exec` depends on `k8s.io/client-go/rest`, so therefore `k8s.io/client-go/tools/auth/exec` already has a dependency on `k8s.io/client-go/plugin/pkg/client/auth/exec`. As long as you are OK with me importing `k8s.io/client-go/plugin/pkg/client/auth/exec` into `k8s.io/client-go/tools/auth/exec`, I'll do that.

[0] https://github.com/kubernetes/kubernetes/pull/95489#discussion_r513538814

Reply:
> I'm just now realizing that k8s.io/client-go/rest depends on k8s.io/client-go/plugin/pkg/client/auth/exec, and k8s.io/client-go/tools/auth/exec depends on k8s.io/client-go/rest, so therefore k8s.io/client-go/tools/auth/exec already has a dependency on k8s.io/client-go/plugin/pkg/client/auth/exec.

That was my thinking as well

Reply:
> I realized I was rewriting a lot of kubeconfig loading logic

Really? I expected a pretty straight copy of the fields… there's no user or context info. It seems like it would be less work and more straightforward than the field by field copy you are doing here

Reply:
I'll go with the straight copy of the fields. Could certainly have been that I was making up stories in my head...

Reply:
Oh nooooo...I think if we followed through with this plan, `plugin/pkg/client/auth/exec` would depend on `rest` (because of return type below), which already depends on `plugin/pkg/client/auth/exec`...

`func ExecClusterToConfig(cluster *clientauthentication.Cluster) (*rest.Config, error)`

Reply:
OK new bad idea: put both `ExecClusterToConfig` and `ConfigToExecCluster` in `rest`, roundtrip test them there, continue to inject the `clientauthentcation.Cluster` into `exec.GetAuthenticator`, and call `ExecClusterToConfig` from `tools/auth/exec`. I think that is the only way the dependency graph will line up.

Reply:
I think you're right. That's not that bad... put them in their own file and there's a certain symmetry to them

Reply:
gave this a shot here: 930bfd161a2